File: python/system_info.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
IP101 System Information Collection Tool

提供跨平台的系统信息收集功能，包括：
- CPU信息（型号、核心数、频率等）
- 内存信息（总内存、可用内存等）
- OpenCV版本信息
- 操作系统信息
"""

import platform
import psutil
import cv2
import subprocess
import sys
import os
from typing import Dict, List, Optional, Tuple
from dataclasses import dataclass


@dataclass
class CPUInfo:
    """CPU信息结构体"""
    vendor: str
    model: str
    cores: int
    threads: int
    frequency_mhz: float
    architecture: str
    features: List[str]


@dataclass
class MemoryInfo:
    """内存信息结构体"""
    total_physical_kb: int
    available_physical_kb: int
    total_virtual_kb: int
    available_virtual_kb: int


@dataclass
class SystemInfoData:
    """系统信息结构体"""
    os_name: str
    os_version: str
    os_architecture: str
    cpu: CPUInfo
    memory: MemoryInfo
    opencv_version: str
    compiler_info: str


class SystemInfo:
    """系统信息收集类"""

    @staticmethod
    def get_system_info() -> SystemInfoData:
        """获取完整的系统信息"""
        return SystemInfoData(
            os_name=SystemInfo.get_os_name(),
            os_version=SystemInfo.get_os_version(),
            os_architecture=SystemInfo.get_architecture(),
            cpu=SystemInfo.get_cpu_info(),
            memory=SystemInfo.get_memory_info(),
            opencv_version=SystemInfo.get_opencv_version(),
            compiler_info=SystemInfo.get_compiler_info()
        )

    @staticmethod
    def get_cpu_info() -> CPUInfo:
        """获取CPU信息"""
        # 获取CPU型号
        model = SystemInfo._get_cpu_model()

        # 获取核心数
        cores = psutil.cpu_count(logical=False)
        threads = psutil.cpu_count(logical=True)

        # 获取CPU频率
        freq = psutil.cpu_freq()
        frequency_mhz = freq.current if freq else 0.0

        # 获取架构
        architecture = SystemInfo.get_architecture()

        # 获取CPU特性
        features = SystemInfo._get_cpu_features()

        return CPUInfo(
            vendor=SystemInfo._get_cpu_vendor(),
            model=model,
            cores=cores,
            threads=threads,
            frequency_mhz=frequency_mhz,
            architecture=architecture,
            features=features
        )

    @staticmethod
    def get_memory_info() -> MemoryInfo:
        """获取内存信息"""
        memory = psutil.virtual_memory()
        swap = psutil.swap_memory()

        return MemoryInfo(
            total_physical_kb=int(memory.total / 1024),
            available_physical_kb=int(memory.available / 1024),
            total_virtual_kb=int((memory.total + swap.total) / 1024),
            available_virtual_kb=int((memory.available + swap.free) / 1024)
        )

    @staticmethod
    def get_opencv_version() -> str:
        """获取OpenCV版本信息"""
        return cv2.__version__

    @staticmethod
    def get_os_name() -> str:
        """获取操作系统名称"""
        return platform.system()

    @staticmethod
    def get_os_version() -> str:
        """获取操作系统版本"""
        return platform.release()

    @staticmethod
    def get_architecture() -> str:
        """获取系统架构"""
        return platform.machine()

    @staticmethod
    def get_compiler_info() -> str:
        """获取编译器信息"""
        return f"Python {sys.version}"

    @staticmethod
    def format_system_info(info: SystemInfoData) -> str:
        """格式化系统信息为字符串"""
        lines = []

        lines.append("=== System Information ===")
        lines.append(f"OS: {info.os_name} {info.os_version} ({info.os_architecture})")
        lines.append(f"Compiler: {info.compiler_info}")
        lines.append(f"OpenCV: {info.opencv_version}")
        lines.append("")

        lines.append("=== CPU Information ===")
        lines.append(f"Model: {info.cpu.model}")
        lines.append(f"Vendor: {info.cpu.vendor}")
        lines.append(f"Cores: {info.cpu.cores} physical, {info.cpu.threads} logical")
        lines.append(f"Frequency: {info.cpu.frequency_mhz:.2f} MHz")
        lines.append(f"Architecture: {info.cpu.architecture}")

        if info.cpu.features:
            lines.append(f"Features: {', '.join(info.cpu.features)}")
        lines.append("")

        lines.append("=== Memory Information ===")
        lines.append(f"Physical Memory: {info.memory.total_physical_kb // 1024 // 1024} GB total, "
                    f"{info.memory.available_physical_kb // 1024 // 1024} GB available")
        lines.append(f"Virtual Memory: {info.memory.total_virtual_kb // 1024 // 1024} GB total, "
                    f"{info.memory.available_virtual_kb // 1024 // 1024} GB available")

        return "\n".join(lines)

    @staticmethod
    def has_cpu_feature(feature: str) -> bool:
        """检查CPU是否支持特定特性"""
        features = SystemInfo._get_cpu_features()
        return feature in features

    @staticmethod
    def _get_cpu_model() -> str:
        """获取CPU型号"""
        if platform.system() == "Windows":
            try:
                result = subprocess.run(
                    ["wmic", "cpu", "get", "name"],
                    capture_output=True, text=True, check=True
                )
                lines = result.stdout.strip().split('\n')
                if len(lines) >= 2:
                    return lines[1].strip()
            except (subprocess.CalledProcessError, FileNotFoundError):
                pass
        elif platform.system() == "Linux":
            try:
                with open('/proc/cpuinfo', 'r') as f:
                    for line in f:
                        if line.startswith('model name'):
                            return line.split(':')[1].strip()
            except FileNotFoundError:
                pass
        elif platform.system() == "Darwin":  # macOS
            try:
                result = subprocess.run(
                    ["sysctl", "-n", "machdep.cpu.brand_string"],
                    capture_output=True, text=True, check=True
                )
                return result.stdout.strip()
            except (subprocess.CalledProcessError, FileNotFoundError):
                pass

        return "Unknown CPU"

    @staticmethod
    def _get_cpu_vendor() -> str:
        """获取CPU厂商"""
        if platform.system() == "Windows":
            try:
                result = subprocess.run(
                    ["wmic", "cpu", "get", "manufacturer"],
                    capture_output=True, text=True, check=True
                )
                lines = result.stdout.strip().split('\n')
                if len(lines) >= 2:
                    return lines[1].strip()
            except (subprocess.CalledProcessError, FileNotFoundError):
                pass
        elif platform.system() == "Linux":
            try:
                with open('/proc/cpuinfo', 'r') as f:
                    for line in f:
                        if line.startswith('vendor_id'):
                            return line.split(':')[1].strip()
            except FileNotFoundError:
                pass
        elif platform.system() == "Darwin":  # macOS
            return "Apple"

        return "Unknown"

    @staticmethod
    def _get_cpu_features() -> List[str]:
        """获取CPU特性"""
        features = []

        if platform.system() == "Linux":
            try:
                with open('/proc/cpuinfo', 'r') as f:
                    content = f.read()
                    if 'mmx' in content.lower():
                        features.append('MMX')
                    if 'sse' in content.lower():
                        features.append('SSE')
                    if 'sse2' in content.lower():
                        features.append('SSE2')
                    if 'sse3' in content.lower():
                        features.append('SSE3')
                    if 'ssse3' in content.lower():
                        features.append('SSSE3')
                    if 'sse4_1' in content.lower():
                        features.append('SSE4.1')
                    if 'sse4_2' in content.lower():
                        features.append('SSE4.2')
                    if 'avx' in content.lower():
                        features.append('AVX')
                    if 'avx2' in content.lower():
                        features.append('AVX2')
            except FileNotFoundError:
                pass
        elif platform.system() == "Darwin":  # macOS
            # macOS下简化处理
            features.extend(['SSE2', 'SSE3', 'SSSE3', 'SSE4.1', 'SSE4.2'])
        elif platform.system() == "Windows":
            # Windows下简化处理
            features.extend(['SSE2', 'SSE3', 'SSSE3', 'SSE4.1', 'SSE4.2'])

        return features


def get_test_environment_info() -> str:
    """获取测试环境信息的便捷函数"""
    """获取测试环境信息

    Returns:
        格式化的测试环境信息字符串
    """
    info = SystemInfo.get_system_info()
    return SystemInfo.format_system_info(info)


def get_image_info(image_path: str) -> str:
    """获取图像信息

    Args:
        image_path: 图像文件路径

    Returns:
        图像信息字符串
    """
    try:
        img = cv2.imread(image_path)
        if img is None:
            return f"Error: Cannot read image {image_path}"

        height, width = img.shape[:2]
        channels = img.shape[2] if len(img.shape) > 2 else 1

        info_lines = []
        info_lines.append(f"Image: {image_path}")
        info_lines.append(f"Size: {width} x {height} pixels")
        info_lines.append(f"Channels: {channels}")
        info_lines.append(f"Data type: {img.dtype}")
        info_lines.append(f"Memory usage: {img.nbytes / 1024:.2f} KB")

        return "\n".join(info_lines)
    except Exception as e:
        return f"Error getting image info: {str(e)}"


if __name__ == "__main__":
    # 示例用法
    print("=== IP101 System Information ===")
    print(get_test_environment_info())

    # 如果有图像文件，显示图像信息
    if len(sys.argv) > 1:
        image_path = sys.argv[1]
        print("\n=== Image Information ===")
        print(get_image_info(image_path))

File: utils/system_info.cpp
#include "system_info.hpp"
#include <sstream>
#include <iomanip>
#include <algorithm>

#ifdef _WIN32
#include <intrin.h>
#elif defined(__linux__)
#include <cpuid.h>
#endif

namespace ip101 {
namespace utils {

SystemInfo::SystemInfoData SystemInfo::getSystemInfo() {
    SystemInfoData info;

    // 获取操作系统信息
    std::string os_info = getOSInfo();
    size_t pos = os_info.find(" ");
    if (pos != std::string::npos) {
        info.os_name = os_info.substr(0, pos);
        info.os_version = os_info.substr(pos + 1);
    } else {
        info.os_name = os_info;
        info.os_version = "Unknown";
    }

    info.os_architecture = getCPUArchitecture();
    info.cpu = getCPUInfo();
    info.memory = getMemoryInfo();
    info.opencv_version = getOpenCVVersion();
    info.compiler_info = getCompilerInfo();

    return info;
}

SystemInfo::CPUInfo SystemInfo::getCPUInfo() {
    CPUInfo cpu;

#ifdef _WIN32
    cpu.model = getWindowsCPUModel();
    cpu.vendor = "Intel"; // 简化处理

    SYSTEM_INFO sysInfo;
    GetSystemInfo(&sysInfo);
    cpu.cores = sysInfo.dwNumberOfProcessors;
    cpu.threads = sysInfo.dwNumberOfProcessors;

    // 获取CPU频率
    HKEY hKey;
    if (RegOpenKeyExA(HKEY_LOCAL_MACHINE,
        "HARDWARE\\DESCRIPTION\\System\\CentralProcessor\\0",
        0, KEY_READ, &hKey) == ERROR_SUCCESS) {
        DWORD mhz;
        DWORD size = sizeof(DWORD);
        if (RegQueryValueExA(hKey, "~MHz", NULL, NULL, (LPBYTE)&mhz, &size) == ERROR_SUCCESS) {
            cpu.frequency_mhz = static_cast<double>(mhz);
        }
        RegCloseKey(hKey);
    }

#elif defined(__linux__)
    cpu.model = getLinuxCPUModel();
    cpu.vendor = "Unknown";

    // 获取核心数
    cpu.cores = sysconf(_SC_NPROCESSORS_ONLN);
    cpu.threads = sysconf(_SC_NPROCESSORS_ONLN);

    // 获取CPU频率
    std::string freq_file = "/sys/devices/system/cpu/cpu0/cpufreq/scaling_cur_freq";
    std::string freq_str = readFileContent(freq_file);
    if (!freq_str.empty()) {
        cpu.frequency_mhz = std::stod(freq_str) / 1000.0;
    }

#elif defined(__APPLE__)
    cpu.model = getMacOSCPUModel();
    cpu.vendor = "Apple";

    // 获取核心数
    int cores;
    size_t size = sizeof(cores);
    if (sysctlbyname("hw.ncpu", &cores, &size, NULL, 0) == 0) {
        cpu.cores = cores;
        cpu.threads = cores;
    }

    // 获取CPU频率
    uint64_t freq;
    size = sizeof(freq);
    if (sysctlbyname("hw.cpufrequency", &freq, &size, NULL, 0) == 0) {
        cpu.frequency_mhz = static_cast<double>(freq) / 1000000.0;
    }
#endif

    cpu.architecture = getCPUArchitecture();
    cpu.features = getCPUFeatures();

    return cpu;
}

SystemInfo::MemoryInfo SystemInfo::getMemoryInfo() {
    MemoryInfo memory;

#ifdef _WIN32
    MEMORYSTATUSEX memInfo;
    memInfo.dwLength = sizeof(MEMORYSTATUSEX);
    if (GlobalMemoryStatusEx(&memInfo)) {
        memory.total_physical_kb = memInfo.ullTotalPhys / 1024;
        memory.available_physical_kb = memInfo.ullAvailPhys / 1024;
        memory.total_virtual_kb = memInfo.ullTotalVirtual / 1024;
        memory.available_virtual_kb = memInfo.ullAvailVirtual / 1024;
    }

#elif defined(__linux__)
    struct sysinfo si;
    if (sysinfo(&si) == 0) {
        memory.total_physical_kb = si.totalram * si.mem_unit / 1024;
        memory.available_physical_kb = si.freeram * si.mem_unit / 1024;
        memory.total_virtual_kb = (si.totalram + si.totalswap) * si.mem_unit / 1024;
        memory.available_virtual_kb = (si.freeram + si.freeswap) * si.mem_unit / 1024;
    }

#elif defined(__APPLE__)
    vm_statistics64_data_t vm_stats;
    mach_msg_type_number_t count = sizeof(vm_stats) / sizeof(natural_t);
    host_t host = mach_host_self();

    if (host_statistics64(host, HOST_VM_INFO64, (host_info64_t)&vm_stats, &count) == KERN_SUCCESS) {
        uint64_t page_size = vm_page_size;
        memory.total_physical_kb = (vm_stats.active_count + vm_stats.inactive_count +
                                   vm_stats.wire_count + vm_stats.free_count) * page_size / 1024;
        memory.available_physical_kb = vm_stats.free_count * page_size / 1024;
        memory.total_virtual_kb = memory.total_physical_kb; // 简化处理
        memory.available_virtual_kb = memory.available_physical_kb; // 简化处理
    }
#endif

    return memory;
}

std::string SystemInfo::getOpenCVVersion() {
    return cv::getVersionString();
}

std::string SystemInfo::getOSInfo() {
#ifdef _WIN32
    return getWindowsOSVersion();
#elif defined(__linux__)
    return getLinuxOSVersion();
#elif defined(__APPLE__)
    return getMacOSVersion();
#else
    return "Unknown OS";
#endif
}

std::string SystemInfo::getCompilerInfo() {
    std::stringstream ss;

#ifdef _MSC_VER
    ss << "MSVC " << _MSC_VER;
#elif defined(__GNUC__)
    ss << "GCC " << __GNUC__ << "." << __GNUC_MINOR__ << "." << __GNUC_PATCHLEVEL__;
#elif defined(__clang__)
    ss << "Clang " << __clang_major__ << "." << __clang_minor__ << "." << __clang_patchlevel__;
#else
    ss << "Unknown Compiler";
#endif

    return ss.str();
}

std::string SystemInfo::formatSystemInfo(const SystemInfoData& info) {
    std::stringstream ss;

    ss << "=== System Information ===" << std::endl;
    ss << "OS: " << info.os_name << " " << info.os_version << " (" << info.os_architecture << ")" << std::endl;
    ss << "Compiler: " << info.compiler_info << std::endl;
    ss << "OpenCV: " << info.opencv_version << std::endl;
    ss << std::endl;

    ss << "=== CPU Information ===" << std::endl;
    ss << "Model: " << info.cpu.model << std::endl;
    ss << "Vendor: " << info.cpu.vendor << std::endl;
    ss << "Cores: " << info.cpu.cores << " physical, " << info.cpu.threads << " logical" << std::endl;
    ss << "Frequency: " << std::fixed << std::setprecision(2) << info.cpu.frequency_mhz << " MHz" << std::endl;
    ss << "Architecture: " << info.cpu.architecture << std::endl;

    if (!info.cpu.features.empty()) {
        ss << "Features: ";
        for (size_t i = 0; i < info.cpu.features.size(); ++i) {
            if (i > 0) ss << ", ";
            ss << info.cpu.features[i];
        }
        ss << std::endl;
    }
    ss << std::endl;

    ss << "=== Memory Information ===" << std::endl;
    ss << "Physical Memory: " << (info.memory.total_physical_kb / 1024 / 1024) << " GB total, "
       << (info.memory.available_physical_kb / 1024 / 1024) << " GB available" << std::endl;
    ss << "Virtual Memory: " << (info.memory.total_virtual_kb / 1024 / 1024) << " GB total, "
       << (info.memory.available_virtual_kb / 1024 / 1024) << " GB available" << std::endl;

    return ss.str();
}

bool SystemInfo::hasCPUFeature(const std::string& feature) {
    auto features = getCPUFeatures();
    return std::find(features.begin(), features.end(), feature) != features.end();
}

std::string SystemInfo::getCPUArchitecture() {
#ifdef _WIN32
#ifdef _WIN64
    return "x86_64";
#else
    return "x86";
#endif
#elif defined(__linux__)
#ifdef __x86_64__
    return "x86_64";
#elif defined(__aarch64__)
    return "ARM64";
#elif defined(__arm__)
    return "ARM32";
#else
    return "Unknown";
#endif
#elif defined(__APPLE__)
#ifdef __x86_64__
    return "x86_64";
#elif defined(__aarch64__)
    return "ARM64";
#else
    return "Unknown";
#endif
#else
    return "Unknown";
#endif
}

std::vector<std::string> SystemInfo::getCPUFeatures() {
    std::vector<std::string> features;

#ifdef _WIN32
    int cpuInfo[4];
    __cpuid(cpuInfo, 1);

    if (cpuInfo[3] & (1 << 23)) features.push_back("MMX");
    if (cpuInfo[3] & (1 << 25)) features.push_back("SSE");
    if (cpuInfo[3] & (1 << 26)) features.push_back("SSE2");
    if (cpuInfo[2] & (1 << 0))  features.push_back("SSE3");
    if (cpuInfo[2] & (1 << 9))  features.push_back("SSSE3");
    if (cpuInfo[2] & (1 << 19)) features.push_back("SSE4.1");
    if (cpuInfo[2] & (1 << 20)) features.push_back("SSE4.2");
    if (cpuInfo[2] & (1 << 28)) features.push_back("AVX");

    // 检查AVX2
    __cpuid(cpuInfo, 7);
    if (cpuInfo[1] & (1 << 5)) features.push_back("AVX2");

#elif defined(__linux__)
    unsigned int eax, ebx, ecx, edx;

    if (__get_cpuid(1, &eax, &ebx, &ecx, &edx)) {
        if (edx & (1 << 23)) features.push_back("MMX");
        if (edx & (1 << 25)) features.push_back("SSE");
        if (edx & (1 << 26)) features.push_back("SSE2");
        if (ecx & (1 << 0))  features.push_back("SSE3");
        if (ecx & (1 << 9))  features.push_back("SSSE3");
        if (ecx & (1 << 19)) features.push_back("SSE4.1");
        if (ecx & (1 << 20)) features.push_back("SSE4.2");
        if (ecx & (1 << 28)) features.push_back("AVX");
    }

    if (__get_cpuid(7, &eax, &ebx, &ecx, &edx)) {
        if (ebx & (1 << 5)) features.push_back("AVX2");
    }

#elif defined(__APPLE__)
    // macOS下简化处理
    features.push_back("SSE2");
    features.push_back("SSE3");
    features.push_back("SSSE3");
    features.push_back("SSE4.1");
    features.push_back("SSE4.2");
#endif

    return features;
}

#ifdef _WIN32

std::string SystemInfo::getWindowsCPUModel() {
    HKEY hKey;
    char buffer[256];
    DWORD size = sizeof(buffer);

    if (RegOpenKeyExA(HKEY_LOCAL_MACHINE,
        "HARDWARE\\DESCRIPTION\\System\\CentralProcessor\\0",
        0, KEY_READ, &hKey) == ERROR_SUCCESS) {
        if (RegQueryValueExA(hKey, "ProcessorNameString", NULL, NULL,
            (LPBYTE)buffer, &size) == ERROR_SUCCESS) {
            RegCloseKey(hKey);
            return std::string(buffer);
        }
        RegCloseKey(hKey);
    }

    return "Unknown CPU";
}

std::string SystemInfo::getWindowsOSVersion() {
    OSVERSIONINFOEXA osvi;
    ZeroMemory(&osvi, sizeof(OSVERSIONINFOEXA));
    osvi.dwOSVersionInfoSize = sizeof(OSVERSIONINFOEXA);

    if (GetVersionExA((OSVERSIONINFOA*)&osvi)) {
        std::stringstream ss;
        ss << "Windows " << osvi.dwMajorVersion << "." << osvi.dwMinorVersion;
        if (osvi.dwBuildNumber > 0) {
            ss << " (Build " << osvi.dwBuildNumber << ")";
        }
        return ss.str();
    }

    return "Windows Unknown";
}

#elif defined(__linux__)

std::string SystemInfo::getLinuxCPUModel() {
    std::string model = readFileContent("/proc/cpuinfo");
    std::istringstream iss(model);
    std::string line;

    while (std::getline(iss, line)) {
        if (line.substr(0, 10) == "model name") {
            size_t pos = line.find(": ");
            if (pos != std::string::npos) {
                return line.substr(pos + 2);
            }
        }
    }

    return "Unknown CPU";
}

std::string SystemInfo::getLinuxOSVersion() {
    std::string os_info = readFileContent("/etc/os-release");
    std::istringstream iss(os_info);
    std::string line;
    std::string name, version;

    while (std::getline(iss, line)) {
        if (line.substr(0, 4) == "NAME") {
            size_t pos = line.find("=");
            if (pos != std::string::npos) {
                name = line.substr(pos + 2);
                name = name.substr(0, name.length() - 1); // 移除引号
            }
        } else if (line.substr(0, 7) == "VERSION") {
            size_t pos = line.find("=");
            if (pos != std::string::npos) {
                version = line.substr(pos + 2);
                version = version.substr(0, version.length() - 1); // 移除引号
            }
        }
    }

    if (!name.empty() && !version.empty()) {
        return name + " " + version;
    }

    return "Linux Unknown";
}

std::string SystemInfo::readFileContent(const std::string& filename) {
    std::ifstream file(filename);
    if (file.is_open()) {
        std::string content((std::istreambuf_iterator<char>(file)),
                           std::istreambuf_iterator<char>());
        file.close();
        return content;
    }
    return "";
}

#elif defined(__APPLE__)

std::string SystemInfo::getMacOSCPUModel() {
    char buffer[256];
    size_t size = sizeof(buffer);

    if (sysctlbyname("machdep.cpu.brand_string", buffer, &size, NULL, 0) == 0) {
        return std::string(buffer);
    }

    return "Unknown CPU";
}

std::string SystemInfo::getMacOSVersion() {
    char buffer[256];
    size_t size = sizeof(buffer);

    if (sysctlbyname("kern.osrelease", buffer, &size, NULL, 0) == 0) {
        return std::string("macOS ") + buffer;
    }

    return "macOS Unknown";
}

#endif

} // namespace utils
} // namespace ip101
